src/databricks/labs/dqx/engine.py

```diff
@@ -535,23 +612,50 @@ def apply_checks_by_metadata_and_save_in_table(
             input_config: Input configuration (e.g., table/view or file location and read options).
             output_config: Output configuration (e.g., table name, mode, and write options).
             quarantine_config: Optional configuration for writing invalid records.
+            metrics_config: Optional configuration for writing summary metrics.
             custom_check_functions: Optional mapping of custom check function names
                 to callables/modules (e.g., globals()).
             ref_dfs: Optional reference DataFrames used by checks.
         """
         # Read data from the specified table
         df = read_input_data(self.spark, input_config)
 
+        if metrics_config and isinstance(self._engine, DQEngineCore) and self._engine.observer and df.isStreaming:
+            listener = self._get_streaming_metrics_listener(
+                input_config=input_config,
+                output_config=output_config,
+                quarantine_config=quarantine_config,
+                metrics_config=metrics_config,
+            )
+            self.spark.streams.addListener(listener)
+
         if quarantine_config:
             # Split data into good and bad records
-            good_df, bad_df = self.apply_checks_by_metadata_and_split(df, checks, custom_check_functions, ref_dfs)
+            good_df, bad_df, observation = self.apply_checks_by_metadata_and_split(
+                df, checks, custom_check_functions, ref_dfs
+            )
             save_dataframe_as_table(good_df, output_config)
             save_dataframe_as_table(bad_df, quarantine_config)
         else:
             # Apply checks and write all data to single table
-            checked_df = self.apply_checks_by_metadata(df, checks, custom_check_functions, ref_dfs)
+            checked_df, observation = self.apply_checks_by_metadata(df, checks, custom_check_functions, ref_dfs)
             save_dataframe_as_table(checked_df, output_config)
 
+        if metrics_config and not df.isStreaming and isinstance(self._engine, DQEngineCore):
```

Comment:
i guess we could encapsulate this in a common private method, as we have the same code in multiple places